### PR TITLE
Task3 duty

### PR DESCRIPTION
**Pull request by TamirOffen**

### CODE REVIEW

File: src/schemas/duty-schemas.js [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+const nameSchema = z.string().min(3).max(50);
+const locationSchema = z.tuple([z.number().min(-90).max(90), z.number().min(-180).max(180)]);
+const rankSchema = z.number().min(0).max(6);
+
+const datetimeSchema = z.union([
+	z.instanceof(Date),
+	z.string().refine((value) => !Number.isNaN(Date.parse(value)), {
+		message: "Invalid date string",
+	}),
+]);
```
Comment (moderj) [inline src/schemas/duty-schemas.js]: read about `.coerce`

File: src/schemas/duty-schemas.js [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+		createdAt: datetimeSchema,
+		updatedAt: datetimeSchema,
+	})
+	.strict();
+
+const messageSchema = z.object({
+	message: z.string(),
+});
+
+const postDutySchema = {
```
Comment (moderj) [inline src/schemas/duty-schemas.js:47]: looks like you are taking a lot of things from the `dutySchema`, so either `.extend` or `.pick`, but always `DRY`

File: src/routes/duty-routes.js [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+export async function dutyRoutes(fastify) {
+	fastify.post("/", { schema: postDutySchema }, async (request, reply) => {
+		try {
+			const newDuty = createDuty(request.body);
+			await fastify.mongo.db.collection("duties").insertOne(newDuty);
+			fastify.log.info({ duty: newDuty }, "Duty created successfully");
+
+			return reply.code(201).send(newDuty);
+		} catch (err) {
+			return reply.status(404).send({ message: err.message });
```
Comment (moderj) [inline src/routes/duty-routes.js]: why `4o4`
why are you not consistent with the `soldier-routes`

File: src/schemas/duty-schemas.js [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+	querystring: z
+		.object({
+			name: nameSchema,
+			description: z.string(),
+			location: locationSchema,
+			startTime: datetimeSchema,
+			endTime: datetimeSchema,
+			minRank: rankSchema,
+			maxRank: rankSchema,
+			constraints: z.string(),
```
Comment (moderj) [inline src/schemas/duty-schemas.js:77]: pretty sure you can convert it into an array if you prefer

File: src/routes/duty-routes.js [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+		const { constraints, ...otherProps } = request.query;
+		const filter = otherProps;
+		if (constraints?.length) filter.constraints = { $all: constraints.split(",") };
+		fastify.log.info({ filter }, "Searching for duties by query");
+
+		const duties =
+			Object.keys(filter).length > 0
+				? await fastify.mongo.db.collection("duties").find(filter).toArray()
+				: [];
+		fastify.log.info({ duties }, "Duties found");
```
Comment (moderj) [inline src/routes/duty-routes.js]: confusing message, when empty, did you find duties?

File: src/routes/duty-routes.js [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+	postDutySchema,
+	putConstraintsSchema,
+} from "../schemas/duty-schemas.js";
+
+export async function dutyRoutes(fastify) {
+	fastify.post("/", { schema: postDutySchema }, async (request, reply) => {
+		try {
+			const newDuty = createDuty(request.body);
+			await fastify.mongo.db.collection("duties").insertOne(newDuty);
+			fastify.log.info({ duty: newDuty }, "Duty created successfully");
```
Comment (moderj) [inline src/routes/duty-routes.js]:
```suggestion
			req.log.info({ duty: newDuty }, "Duty created successfully");
```

big difference between the two, see why!

File: src/routes/duty-routes.js [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+			fastify.log.info({ id }, "Duty not found!");
+			return reply.status(404).send({ message: `Duty not found with id ${id}` });
+		}
+
+		fastify.log.info({ id }, "Duty found");
+
+		return reply.status(200).send(duty);
+	});
+
+	fastify.delete("/:id", { schema: deleteDutySchema }, async (request, reply) => {
```
Comment (moderj) [inline src/routes/duty-routes.js:59]: did you check if it's scheduled?

File: src/routes/duty-routes.js [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+				{ id },
+				`Cannot update duty to have minRank ${duty.minRank} > maxRank ${request.body.maxRank}`,
+			);
+			return reply.status(404).send({
+				message: `Cannot update duty to have minRank ${duty.minRank} > maxRank ${request.body.maxRank}`,
+			});
+		}
+		if (duty?.status === "scheduled") {
+			fastify.log.info({ id }, "Cannot update scheduled duty");
+			return reply.status(404).send({ message: "Cannot update scheduled duty" });
```
Comment (moderj) [inline src/routes/duty-routes.js]: is it 404?

File: src/schemas/duty-schemas.js [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+		200: messageSchema,
+		404: messageSchema,
+	},
+};
+
+const patchDutySchema = {
+	params: z.object({
+		id: ObjectIDStringSchema,
+	}),
+	body: z
```
Comment (moderj) [inline src/schemas/duty-schemas.js:103]: DRY

File: src/schemas/duty-schemas.js [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+		.object({
+			name: nameSchema,
+			description: z.string(),
+			location: locationSchema,
+			startTime: datetimeSchema,
```
Comment (moderj) [inline src/schemas/duty-schemas.js]: did you validate it is in the future?

File: test/duty.test.js [hunk 11]
```diff
@@ -0,0 +1,329 @@
+import { afterEach, beforeAll, beforeEach, describe, expect, it } from "vitest";
+import { generateDuty, generatePostDuty } from "./data-factory.js";
+import { createFastifyApp } from "./src/app.js";
```
Comment (moderj) [inline test/duty.test.js]: fix

File: test/duty.test.js [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+			await fastify.mongo.db.collection("duties").insertMany([duty1, duty2]);
+			const duty1ID = duty1._id.toString();
+
+			const getDutyResponse = await fastify.inject({
+				method: "GET",
+				url: `/duties/${duty1ID}`,
+			});
+
+			expect(getDutyResponse.statusCode).toBe(200);
+			expect(getDutyResponse.json()._id).toBe(duty1ID);
```
Comment (moderj) [inline test/duty.test.js]: match the response properties to be as you would expect

File: test/duty.test.js [hunk 13]
```diff
@@ -0,0 +1,329 @@
...
+				method: "GET",
+				url: `/duties/${"".padStart(24, "0")}`,
+			});
+
+			expect(dutyDNEResponse.statusCode).toBe(404);
+		});
+	});
+
+	describe("DELETE /duties/:id", () => {
+		it("Should return status 200, if duty exists", async () => {
```
Comment (moderj) [inline test/duty.test.js]: you make the tests, let me know what's the edge case you are testing
`if duty exists` is not how you want to describe your test

File: test/duty.test.js [hunk 14]
```diff
@@ -0,0 +1,329 @@
...
+			const duty1 = generateDuty({});
+			const duty2 = generateDuty({});
+			await fastify.mongo.db.collection("duties").insertMany([duty1, duty2]);
+			const duty1ID = duty1._id.toString();
+
+			const deleteDutyResponse = await fastify.inject({
+				method: "DELETE",
+				url: `/duties/${duty1ID}`,
+			});
+			expect(deleteDutyResponse.statusCode).toBe(200);
```
Comment (moderj) [inline test/duty.test.js]: you've checked it returned 200, good.
is it supposed to delete it from the database?

File: test/duty.test.js [hunk 15]
```diff
@@ -0,0 +1,329 @@
...
+
+			const patchResponse = await fastify.inject({
+				method: "PATCH",
+				url: `/duties/${originalDuty._id.toString()}`,
+				payload: updateToDuty,
+			});
+			const updatedDuty = patchResponse.json();
+
+			expect(patchResponse.statusCode).toBe(200);
+			for (const key in updateToDuty) expect(updatedDuty[key]).toStrictEqual(updateToDuty[key]);
```
Comment (moderj) [inline test/duty.test.js]: find another way to match an object

File: test/duty.test.js [hunk 16]
```diff
@@ -0,0 +1,329 @@
...
+		});
+
+		it("Should return status 404, if ID DNE", async () => {
```
Comment (moderj) [inline test/duty.test.js]: just write do not exist

File: test/duty.test.js [hunk 17]
```diff
@@ -0,0 +1,329 @@
...
+				payload: newPropertyUpdate,
+			});
+
+			expect(patchResponse.statusCode).toBe(400);
+			for (const newProp of Object.keys(newPropertyUpdate)) {
+				expect(originalDuty).not.toHaveProperty(newProp);
+			}
+		});
+
+		it("Should not be able to update duty to have minRank > maxRank", async () => {
```
Comment (moderj) [inline test/duty.test.js]: what if minRank is saved as `{ minRank: 5 }` and you try to update PATCH with `{ maxRank: 3 }`

File: src/models/duty.js [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+		location: dutyParams.location,
+		startTime: new Date(dutyParams.startTime),
+		endTime: new Date(dutyParams.endTime),
+		constraints: dutyParams.constraints,
+		soldiersRequired: dutyParams.soldiersRequired,
+		value: dutyParams.value,
+		soldiers: [],
+		status: dutyParams.status || "unscheduled",
+		createdAt: new Date(),
+		updatedAt: new Date(),
```
Comment (moderj) [inline src/models/duty.js]: make it the same `new Date()`

File: src/models/duty.js [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+		name: dutyParams.name,
+		description: dutyParams.description,
+		location: dutyParams.location,
+		startTime: new Date(dutyParams.startTime),
+		endTime: new Date(dutyParams.endTime),
+		constraints: dutyParams.constraints,
+		soldiersRequired: dutyParams.soldiersRequired,
+		value: dutyParams.value,
+		soldiers: [],
+		status: dutyParams.status || "unscheduled",
```
Comment (moderj) [inline src/models/duty.js]: you should not be provided with `dutyParams.status`

File: src/models/duty.js [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+		newDuty.maxRank = dutyParams.maxRank;
+	}
+	if (dutyParams.minRank && dutyParams.maxRank && dutyParams.minRank > dutyParams.maxRank) {
+		throw new Error("minRank cannot be bigger than maxRank");
+	}
+
+	return newDuty;
+}
+
+export { createDuty };
```
Comment (moderj) [inline src/models/duty.js]: just export where you declare the function

File: test/duty.test.js [hunk 21]
```diff
@@ -0,0 +1,329 @@
...
+			const newPropertyUpdate = { new_property: "property" };
+
+			const patchResponse = await fastify.inject({
+				method: "PATCH",
+				url: `/duties/${originalDuty._id.toString()}`,
+				payload: newPropertyUpdate,
+			});
+
+			expect(patchResponse.statusCode).toBe(400);
+			for (const newProp of Object.keys(newPropertyUpdate)) {
```
Comment (moderj) [inline test/duty.test.js]: same as the last `for` comment, try to avoid this

File: test/duty.test.js (hunk 17 above)
Comment (TamirOffen) [reply test/duty.test.js]: it should not be able to update the duty, and return status code 400

File: src/routes/duty-routes.js [hunk 22]
```diff
@@ -0,0 +1,137 @@
...
+export async function dutyRoutes(fastify) {
+	fastify.post("/", { schema: postDutySchema }, async (request, reply) => {
+		try {
+			const newDuty = createDuty(request.body);
+			await fastify.mongo.db.collection("duties").insertOne(newDuty);
+			request.log.info({ duty: newDuty }, "Duty created successfully");
+
+			return reply.code(201).send(newDuty);
+		} catch (err) {
+			return reply.status(400).send({ message: err.message });
```
Comment (moderj) [inline src/routes/duty-routes.js:21]: should it always be `400`?
why do you handle it differently than the `POST /soldiers` route

File: src/schemas/duty-schemas.js [hunk 23]
```diff
@@ -0,0 +1,151 @@
...
+		201: dutySchema,
+		404: messageSchema,
+	},
+};
+
+const getDutyByQuerySchema = {
+	querystring: dutySchema
+		.omit({
+			_id: true,
+			constraint: true,
```
Comment (moderj) [inline src/schemas/duty-schemas.js:70]:
```suggestion
			constraints: true,
```

File: src/schemas/duty-schemas.js [hunk 24]
```diff
@@ -0,0 +1,151 @@
...
+		.partial()
+		.refine(
+			(data) => {
+				if (data.startTime && data.endTime) {
+					return data.endTime > data.startTime;
+				}
+				return true;
+			},
+			{ message: "endTime must be after startTime" },
+		),
```
Comment (moderj) [inline src/schemas/duty-schemas.js:126]:
```suggestion
		.refine((data) => (data.startTime && data.endTime ? data.endTime > data.startTime : true), {
			message: "endTime must be after startTime",
		}),
```

File: test/data-factory.js [hunk 25]
```diff
@@ -1,3 +1,5 @@
+import { date } from "zod";
```
Comment (moderj) [inline test/data-factory.js:1]: remove

File: test/data-factory.js [hunk 26]
```diff
@@ -1,3 +1,5 @@
...
+import { createDuty } from "../src/models/duty";
 import { getSoldierRank } from "../src/models/soldier";
 
 function generatePostSoldier(soldierParams) {
```
Comment (moderj) [inline test/data-factory.js:5]:
```suggestion
function generatePostSoldier(soldierParams = {}) {
```

default to empty object in all generators, so you don't need to explicitly pass an empty object in your tests

File: test/data-factory.js [hunk 27]
```diff
@@ -44,4 +46,41 @@ function generateSoldier(soldierParams) {
...
+		...(status && { status }),
+		...(minRank && { minRank }),
+		...(maxRank && { maxRank }),
+	};
+
+	return newDuty;
+}
+
+function generateDuty(dutyParams) {
+	return createDuty(generatePostDuty(dutyParams));
```
Comment (moderj) [inline test/data-factory.js:83]: be consistent with the way you create the data, either do the same with `generateSoldier` or change the `generateDuty` to match the other one.

File: test/duty.test.js [hunk 28]
```diff
@@ -0,0 +1,400 @@
...
+				constraints: ["Night duty", "Perimeter security"],
+				value: 1500.5,
+				minRank: 2,
+			};
+			const response = await fastify.inject({
+				method: "POST",
+				url: "/duties",
+				payload: missingPropsDuty,
+			});
+			expect(response.statusCode).toBe(400);
```
Comment (moderj) [inline test/duty.test.js:52]: empty line before `expect`

File: test/duty.test.js [hunk 29]
```diff
@@ -0,0 +1,400 @@
...
+			expect(response.statusCode).toBe(201);
+			expect(returnedDuty).toMatchObject(dutyPost);
+			expect(returnedDuty._id).toBeDefined();
+			expect(returnedDuty.createdAt).toBeDefined();
+			expect(returnedDuty.updatedAt).toBeDefined();
+			expect(returnedDuty.statusHistory).toBeDefined();
+		});
+
+		it("Add duty with missing properties should return status 400", async () => {
+			const missingPropsDuty = {
```
Comment (moderj) [inline test/duty.test.js:39]: don't manually create it, use `generatePostDuty` and use `destructure` to omit one or more of its properties

File: test/duty.test.js [hunk 30]
```diff
@@ -0,0 +1,400 @@
...
+			};
+			const response = await fastify.inject({
+				method: "POST",
+				url: "/duties",
+				payload: missingPropsDuty,
+			});
+			expect(response.statusCode).toBe(400);
+		});
+
+		it("Add duty duty with illegal GeoJSON location should return status 400", async () => {
```
Comment (moderj) [inline test/duty.test.js:55]: try to make the test description readable with the `it` prefix.
perhaps consider starting always with `it should...`

File: test/duty.test.js [hunk 31]
```diff
@@ -0,0 +1,400 @@
...
+			const response = await fastify.inject({
+				method: "POST",
+				url: "/duties",
+				payload: generatePostDuty({ startTime: earlyStartTime }),
+			});
+			expect(response.statusCode).toBe(400);
+		});
+	});
+
+	describe("GET /duties?", () => {
```
Comment (moderj) [inline test/duty.test.js:98]: add test for when not sending querystring at all

File: test/duty.test.js [hunk 32]
```diff
@@ -0,0 +1,400 @@
...
+			const noDutiesResponse = await fastify.inject({
+				method: "GET",
+				url: `/duties?name=${nonExistentDutyName}`,
+			});
+
+			expect(noDutiesResponse.json().length).toBe(0);
+		});
+	});
+
+	describe("GET /duties/:id", () => {
```
Comment (moderj) [inline test/duty.test.js:158]: what happens when given id that does not match the `ObjectId` pattern?

File: test/duty.test.js [hunk 33]
```diff
@@ -0,0 +1,400 @@
...
+			const dutyDNEResponse = await fastify.inject({
+				method: "DELETE",
+				url: `/duties/${"".padStart(24, "0")}`,
+			});
+			expect(dutyDNEResponse.statusCode).toBe(404);
+		});
+
+		it("Cannot not delete a scheduled duty, and should return status code 400", async () => {
+			const duty = generateDuty({});
+			duty.status = "scheduled";
```
Comment (moderj) [inline test/duty.test.js:221]: why not generating it with the wanted status?

File: test/duty.test.js [hunk 34]
```diff
@@ -0,0 +1,400 @@
...
+			const duty1 = generateDuty({});
+			const duty2 = generateDuty({});
+			await fastify.mongo.db.collection("duties").insertMany([duty1, duty2]);
+			const duty1ID = duty1._id.toString();
+
+			const deleteDutyResponse = await fastify.inject({
+				method: "DELETE",
+				url: `/duties/${duty1ID}`,
+			});
+			const getDutyResponse = await fastify.inject({
```
Comment (moderj) [inline test/duty.test.js:202]: replace it with a simple db query

File: test/duty.test.js [hunk 35]
```diff
@@ -0,0 +1,400 @@
...
+			const duty = generateDuty({});
+			duty.status = "scheduled";
+			await fastify.mongo.db.collection("duties").insertOne(duty);
+			const dutyID = duty._id.toString();
+
+			const dutyDNEResponse = await fastify.inject({
+				method: "DELETE",
+				url: `/duties/${dutyID}`,
+			});
+			const getDutyResponse = await fastify.inject({
```
Comment (moderj) [inline test/duty.test.js:229]: same

File: test/duty.test.js [hunk 36]
```diff
@@ -0,0 +1,400 @@
...
+				url: `/duties/${"".padStart(24, "0")}`,
+				payload: updateToDuty,
+			});
+
+			expect(updateDutyResponse.statusCode).toBe(404);
+		});
+
+		it("Cannot update scheduled duties, should return status 400", async () => {
+			const scheduledDuty = generateDuty({});
+			scheduledDuty.status = "scheduled";
```
Comment (moderj) [inline test/duty.test.js:279]: same

File: test/duty.test.js [hunk 37]
```diff
@@ -0,0 +1,400 @@
...
+			const putDutyResponse = await fastify.inject({
+				method: "PUT",
+				url: `/duties/${"".padStart(24, "0")}/constraints`,
+				payload: [],
+			});
+
+			expect(putDutyResponse.statusCode).toBe(404);
+		});
+
+		it("Cannot not have duplicate constraints", async () => {
```
Comment (moderj) [inline test/duty.test.js:380]: fix message, and add status code...
if it cannot have duplicates then how does it return 200?

File: test/duty.test.js [hunk 38]
```diff
@@ -0,0 +1,400 @@
...
+			const originalConstraints = ["Aleph", "Bet"];
+			const newConstraints = ["Aleph", "Bet", "C"];
```
Comment (moderj) [inline test/duty.test.js:382]:
```suggestion
			const newConstraints = originalConstraints.concat(["C"]);
```

File: src/routes/duty-routes.js [hunk 39]
```diff
@@ -0,0 +1,137 @@
...
+		const filter = otherProps;
+		if (constraints?.length) filter.constraints = { $all: constraints.split(",") };
+		request.log.info({ filter }, "Searching for duties by query");
+
+		const duties =
+			Object.keys(filter).length > 0
+				? await fastify.mongo.db.collection("duties").find(filter).toArray()
+				: [];
+		if (duties.length) request.log.info("No duties found");
+		else request.log.info({ duties }, "Duties found");
```
Comment (moderj) [inline src/routes/duty-routes.js:36]: you wouldn't want to log the entire duties array, as it's too much data.
you can log the ids, and the number of duties found
